### ISSUE TITLE
API checker needs to support multi-targeted projects

### ISSUE THREAD

**Issue by jaredpar** [dotnet/roslyn-analyzers#3187]
Roslyn is [intending to change](https://github.com/dotnet/roslyn/issues/40766) many of the core projects, including those which have public APIs, to be multi-targeted. That potentially leads to an issue with the API checker as it will be asked to validate the public API surface area twice: once for netstandard2.0 and netcoreapp3.1.

That will lead to issues if the API differs between the different target frameworks. At the moment there is no intent to deviate here hence the double validation is likely to just be more work, not an actual blocker. The ability to deviate may become important in the future as we may want to expose new APIs / types across the netcoreapp3.1 build chain. At that point we will need to have a way to split up the API across different files based on TF.

**Comment by jaredpar** [comment]
@jcouv @chsienki something for us to think about. I don't have a great answer yet but feels like a problem we will eventually need to solve.

**Comment by mavasani** [comment]
Adding some more context.
This was requested in https://github.com/dotnet/roslyn-analyzers/issues/2621. @drewnoakes added documentation on how clients can potentially handle this with current support (separate public API files and validation per TF): https://github.com/dotnet/roslyn-analyzers/pull/3046.

**Comment by roji** [comment]
Another point here... In projects which multi-target, the actual API divergence is typically quite small, but the current method of handling multi-targeting is to duplicate the entire shipped/unshipped files. For a project targeting 4 TFMs, that means that most public API changes need to be manually duplicated 4 times.

Ideally, the analyzer would allow having one shipped/unshipped fileset at the project root, and merge that with a TFM-conditional fileset; thus all common APIs would be in one place, and TFM-specific ones in another. Let me know if you'd prefer another issue for this.

**Comment by NickCraver** [comment]
A fun example of this in latest is fighting analyzers, for example:
![image](https://user-images.githubusercontent.com/454813/123546816-402a9f00-d72c-11eb-8a6c-036e499be831.png)
![image](https://user-images.githubusercontent.com/454813/123546818-428cf900-d72c-11eb-9d98-7feb115d157d.png)

The net5.0 analyzers are largely compatible with previous frameworks, but some analyzers on what's needed, what should or shouldn't be suppressed, etc. can be dueling it out. Applying a fix for one framework will break another.

**Comment by twsouthwick** [comment]
I'm hitting this with a project that has 40k public APIs, with 5 different TFMs. There is a delta of only about 5 APIs difference between the TFMs, so this ends up adding about 20mb to my project for all the TFMS.

Could the analyzer be updated to handle multiple `PublicAPI.*.txt` documents? I've only looked as far as how it's parsed in and see it's only reading a single file; if we were to add a filter for `PublicAPI.Shipped*.txt` and `PublicAPI.Unshipped*.txt`, I could potentially add a `PublicAPI.Shipped.net46.txt` file for some APIs that are exposed there that don't exist elsewhere.

If this behavior is acceptable, I'm happy to provide a PR for it.

**Comment by nohwnd** [comment]
I've ran into the same in vstest and your suggestion @twsouthwick would be awesome to have. 

I would love this to work in a way that simply concatenates all the PublicAPI.Shipped*.txt files that are listed in the proj file (and same for the Unshipped of course). 

This would hopefully allow hierarchy like this to work:
![image](https://user-images.githubusercontent.com/5735905/132499834-52b18a40-eb9d-4db2-a9ba-46650eeef526.png)

```xml
  <ItemGroup>
    <AdditionalFiles Include="PublicAPI\PublicAPI.Shipped.txt" />
    <AdditionalFiles Include="PublicAPI\PublicAPI.Unshipped.txt" />
    <AdditionalFiles Include="PublicAPI\$(TargetFramework)\PublicAPI.Shipped.txt" />
    <AdditionalFiles Include="PublicAPI\$(TargetFramework)\PublicAPI.Unshipped.txt" />
  </ItemGroup>
```

And would also be flexible enough to work for a case where we differ only between netframework and net (core), but build for `net472;net48;` and `netstandard2.0;netstandard2.1;net5.0;net6.0`.

![image](https://user-images.githubusercontent.com/5735905/132500150-133b4f8d-8159-479e-8cbf-75c55b1dcff3.png)
